src/gerbil/runtime/mop.ss

```diff
@@ -6,449 +6,554 @@ package: gerbil/runtime
+;; Compute the flags and field-info and create a type-descriptor
+(def (make-type-descriptor* type-id type-name type-super
+                            precedence-list all-slots alist
+                            constructor slot-table methods)
+  #;(DBG make-type-descriptor*:
+       type-id type-name 'type-super (and type-super (##type-id type-super))
+       (map ##type-id precedence-list) all-slots (alist-form alist)
+       constructor (hash->list slot-table) methods)
+  ;; compute a table of slots with print: or equal: or transparent: flag
+  ;; ht: table to which to add according slots
+  ;; key: either print: or equal: (both implied by transparent:)
   (def (put-props! ht key)
-    (def (put-plist! ht key plist)
-      (cond
-       ((assgetq key plist)
-        => (lambda (lst)
-             (for-each (lambda (id) (hash-put! ht id #t)) lst)))))
-
-    (put-plist! ht key rtd-plist)
-    (when rtd-mixin
-      (for-each (lambda (klass)
-                  (when (type-descriptor-mixin klass) ; ignore structs
-                    (let (plist (type-descriptor-plist klass))
-                      (if (assgetq transparent: plist)
-                        (put-plist! ht slots: plist)
-                        (put-plist! ht key plist)))))
-                rtd-mixin)))
-
-  (let* ((transparent? (assgetq transparent: rtd-plist))
-         (field-names
-          (cond
-           ((assq fields: rtd-plist) => cdr)
-           (else '())))
-         (field-names
-          (cond
-           ((assq slots: rtd-plist)
-            => (lambda (slots)
-                 (append field-names (cdr slots))))
-           (else field-names)))
-         (_
-          (unless (fx= rtd-fields (length field-names))
-            (error "Bad field descriptor; length mismatch" type-id rtd-fields field-names)))
-         (canonical-fields
-          (if type-super
-            (list-tail field-names (type-descriptor-fields type-super))
-            field-names))
+    (def (put-slots! ht slots)
+      (for-each (cut hash-put! ht <> #t) slots))
+    (def (put-alist! ht key alist)
+      (cond ((assgetq key alist) => (cut put-slots! ht <>))))
+    (put-alist! ht key alist)
+    (for-each (lambda (mixin)
+                (let (alist (type-descriptor-alist mixin))
+                  (if (assgetq transparent: alist)
+                    (put-slots! ht (cdr (vector->list (type-descriptor-all-slots mixin))))
+                    (put-alist! ht key alist))))
+              precedence-list))
+
+  (let* ((transparent? (assgetq transparent: alist))
+         (all-slots-printable? (or transparent? (assgetq print: alist)))
          (printable
-          (if transparent?
+          (if all-slots-printable?
             #f                          ; all printable
             (let (ht (make-hash-table-eq))
               (put-props! ht print:)
               ht)))
-         (equality
-          (if transparent?
+         (all-slots-equalable? (or transparent? (assgetq equal: alist)))
```

Comment:
this is subtly wrong? what if equal is a symbol list for which ones to equal compare?

Reply:
Better?

I probably need to add test cases... Sigh.